nipoppy/workflows/pipeline.py

```diff
@@ -400,21 +404,96 @@ def run_main(self):
             analysis_level=self.pipeline_step_config.ANALYSIS_LEVEL,
         )
 
+        if self.hpc:
+            self.submit_hpc_job(participants_sessions)
+        else:
+            # Default behavior
+            for participant_id, session_id in participants_sessions:
+                self.n_total += 1
+                self.logger.info(
+                    f"Running for participant {participant_id}, session {session_id}"
+                )
+                try:
+                    self.run_single(participant_id, session_id)
+                    self.n_success += 1
+                except Exception as exception:
+                    self.return_code = ReturnCode.PARTIAL_SUCCESS
+                    self.logger.error(
+                        f"Error running {self.pipeline_name} {self.pipeline_version}"
+                        f" on participant {participant_id}, session {session_id}"
+                        f": {exception}"
+                    )
+
+    def submit_hpc_job(self, participants_sessions):
+        """Submit jobs to a HPC cluster for processing."""
+        self.logger.info("Running in HPC mode.")
+
+        hpc_templates_path = Path(f"{self.dpath_root}/code/hpc_templates/{self.hpc}")
+        qa = QueueAdapter(directory=str(hpc_templates_path))
+
+        # Generate the list of nipoppy commands as a single string for a shell array
+        job_array_commands = []
         for participant_id, session_id in participants_sessions:
-            self.n_total += 1
+            command = [
+                "nipoppy",
+                "run",
+                str(self.dpath_root),
+                "--pipeline",
+                self.pipeline_name,
+                "--pipeline-version",
+                self.pipeline_version,
+                "--pipeline-step",
+                self.pipeline_step,
+                "--participant-id",
+                participant_id,
+                "--session-id",
+                session_id,
+            ]
+            job_array_commands.append(shlex.join(command))
+        # Join the commands into a single string
+
+        job_array_commands_str = " ".join([f'"{cmd}"' for cmd in job_array_commands])
+        module_load = self.config.HPC_PREAMBLE
+
+        # Build the single command to submit as an array job
+        if self.hpc == "slurm":
+            command = (
+                f"bash -c '{module_load}; commands=({job_array_commands_str}); "
```

Comment:
The -c explicitly states to treat everything inside the quotes as a single string of commands to execute. It guarantees the string is correctly parsed and executed as a script. Being explicit with it avoids ambiguity - both for the human and for the interpreter.

Without -c, we could see different behaviour in different shells - especially if bash is not the default shell. Another shell could treat the commands as arguments or as part of its environment setup, instead of as a list of commands to execute.
eg: I could imagine 'bash commands=(${job_array_commands_str})' invoking a command named bash, with commands=(${job_array_commands_str}) as an argument to it, rather than executing the commands stored in the array. Omitting -c seems like a great way to make future devs spend hours debugging escape quotation marks.

It's currently working, using -c seems a safer choice across shells, and to me it makes sense to leave as is for now. However, if it causes any issues down the line, I'm happy to revisit it and make changes.